Test config with `tuple` output value results in a panic

### terraform-plugin-testing version
<!---
Inspect your go.mod as below to find the version, and paste the result between
the ``` marks below.

go list -m github.com/hashicorp/terraform-plugin-testing/...

If you are not running the latest version of terraform-plugin-testing, please try
upgrading because your bug may have already been fixed.
-->

```
v1.7.0
```

### Actual Behavior
The following test causes a panic when the testing framework attempts to shim the output from `terraform show -json`
```go
func TestRecreateTuplePanic(t *testing.T) {
    resource.UnitTest(t, resource.TestCase{
        ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
            "framework": providerserver.NewProtocol6WithError(New()),
        },
        Steps: []resource.TestStep{
            {
                Config: `
                output "test" {
                    value = [true, "hello tuple!"]
                }`,
            },
        },
    })
}
```

#### Terraform State JSON output
```json
{
  "format_version": "1.0",
  "terraform_version": "1.8.0",
  "values": {
    "outputs": {
      "test": {
        "sensitive": false,
        "value": [
          true,
          "hello tuple!"
        ],
        "type": [
          "tuple",
          [
            "bool",
            "string"
          ]
        ]
      }
    },
    "root_module": {}
  }
}
```

#### Relevant stack trace
```bash
2024-03-22T15:19:40.253-0400 [TRACE] sdk.helper_resource: Called Terraform CLI show command for JSON state: test_working_directory=/var/folders/t8/1tjpvj_d24x8yl0st5qswzf80000gp/T/plugintest66285806 test_step_number=1 test_name=TestRecreateTuplePanic test_terraform_path=/usr/local/bin/terraform
--- FAIL: TestRecreateTuplePanic (0.38s)
panic: interface conversion: interface {} is string, not bool [recovered]
	panic: interface conversion: interface {} is string, not bool

goroutine 35 [running]:
testing.tRunner.func1.2({0x105248400, 0x1400064aed0})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1631 +0x1c4
testing.tRunner.func1()
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1634 +0x33c
panic({0x105248400?, 0x1400064aed0?})
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/panic.go:770 +0x124
github.com/hashicorp/terraform-plugin-testing/helper/resource.shimOutputState(0x1400064a9c0)
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/state_shim.go:77 +0x574
github.com/hashicorp/terraform-plugin-testing/helper/resource.shimStateFromJson(0x140000f23c0)
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/state_shim.go:33 +0xe8
github.com/hashicorp/terraform-plugin-testing/helper/resource.getState({0x105380e90, 0x140003ac7b0}, {0x10538b4a8, 0x140001396c0}, 0x140000f2280)
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing_new.go:471 +0x50
github.com/hashicorp/terraform-plugin-testing/helper/resource.runNewTest.func1.1()
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing_new.go:71 +0x44
github.com/hashicorp/terraform-plugin-testing/helper/resource.runProviderCommand({0x105380e90, 0x140003ac7b0}, {0x10538b4a8, 0x140001396c0}, 0x140002975e8, 0x140000f2280, 0x140002997f8)
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/plugin.go:438 +0x1ddc
github.com/hashicorp/terraform-plugin-testing/helper/resource.runNewTest.func1()
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing_new.go:70 +0xc8
github.com/hashicorp/terraform-plugin-testing/helper/resource.runNewTest({0x105380e90, 0x140003ac7b0}, {0x10538b4a8, 0x140001396c0}, {0x1, 0x0, {0x0, 0x0, 0x0}, 0x0, ...}, ...)
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing_new.go:462 +0x11d0
github.com/hashicorp/terraform-plugin-testing/helper/resource.Test({0x10538b4a8, 0x140001396c0}, {0x1, 0x0, {0x0, 0x0, 0x0}, 0x0, 0x0, 0x140002adb30, ...})
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing.go:922 +0x59c
github.com/hashicorp/terraform-plugin-testing/helper/resource.UnitTest({0x10538b4a8, 0x140001396c0}, {0x1, 0x0, {0x0, 0x0, 0x0}, 0x0, 0x0, 0x140002adb30, ...})
	/Users/austin.valle/go/pkg/mod/github.com/hashicorp/terraform-plugin-testing@v1.7.0/helper/resource/testing.go:936 +0x68
github.com/hashicorp/terraform-provider-corner/internal/framework6provider.TestRecreateTuplePanic(0x140001396c0)
	/Users/austin.valle/code/terraform-provider-corner/internal/framework6provider/dynamic_function_test.go:44 +0x154
testing.tRunner(0x140001396c0, 0x1053702b8)
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 1
	/opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1742 +0x318
FAIL	github.com/hashicorp/terraform-provider-corner/internal/framework6provider	0.902s
FAIL
```

### Expected Behavior

The state shimming logic needs to be updated to support tuple values that might have different types in an array 😅 

https://github.com/hashicorp/terraform-plugin-testing/blob/b5ef730f9b2c245fe2ff5054bd3a63c93b08e695/helper/resource/state_shim.go#L59-L95

### References
- Found while adding dynamic return tests in https://github.com/hashicorp/terraform-provider-corner/pull/228

--- a/.changes/unreleased/BUG FIXES-20240325-125136.yaml
+++ b/.changes/unreleased/BUG FIXES-20240325-125136.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: 'helper/resource: Fix panic in output state shimming when a tuple is present.'
+time: 2024-03-25T12:51:36.841767-04:00
+custom:
+  Issue: "310"

--- a/helper/resource/state_shim.go
+++ b/helper/resource/state_shim.go
@@ -62,27 +62,45 @@ func shimOutputState(so *tfjson.StateOutput) (*terraform.OutputState, error) {
 			os.Value = v
 			return os, nil
 		}
+
 		switch firstElem := v[0].(type) {
 		case string:
 			elements := make([]interface{}, len(v))
 			for i, el := range v {
-				//nolint:forcetypeassert // Guaranteed by type switch
-				elements[i] = el.(string)
+				strElement, ok := el.(string)
+				// If the type of the element doesn't match the first elem, it's a tuple, return the original value
+				if !ok {
+					os.Value = v
+					return os, nil
+				}
+				elements[i] = strElement
 			}
 			os.Value = elements
 		case bool:
 			elements := make([]interface{}, len(v))
 			for i, el := range v {
-				//nolint:forcetypeassert // Guaranteed by type switch
-				elements[i] = el.(bool)
+				boolElement, ok := el.(bool)
+				// If the type of the element doesn't match the first elem, it's a tuple, return the original value
+				if !ok {
+					os.Value = v
+					return os, nil
+				}
+
+				elements[i] = boolElement
 			}
 			os.Value = elements
 		// unmarshalled number from JSON will always be json.Number
 		case json.Number:
 			elements := make([]interface{}, len(v))
 			for i, el := range v {
-				//nolint:forcetypeassert // Guaranteed by type switch
-				elements[i] = el.(json.Number)
+				numberElement, ok := el.(json.Number)
+				// If the type of the element doesn't match the first elem, it's a tuple, return the original value
+				if !ok {
+					os.Value = v
+					return os, nil
+				}
+
+				elements[i] = numberElement
 			}
 			os.Value = elements
 		case []interface{}:

--- a/helper/resource/state_shim_test.go
+++ b/helper/resource/state_shim_test.go
@@ -1,3 +1,6 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
 package resource
 
 import (

--- a/helper/resource/state_shim_test.go
+++ b/helper/resource/state_shim_test.go
@@ -1,0 +1,304 @@
+package resource
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-testing/internal/testing/testprovider"
+	"github.com/hashicorp/terraform-plugin-testing/internal/testing/testsdk/providerserver"
+	"github.com/hashicorp/terraform-plugin-testing/knownvalue"
+	"github.com/hashicorp/terraform-plugin-testing/statecheck"
+)
+
+func TestStateShimOutput_String(t *testing.T) {
+	t.Parallel()
+
+	Test(t, TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"test": providerserver.NewProviderServer(testprovider.Provider{}),
+		},
+		Steps: []TestStep{
+			{
+				Config: `output "test" {
+					value = "hello world"
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.StringExact("hello world"),
+					),
+				},
+			},
+		},
+	})
+}
+
+func TestStateShimOutput_List(t *testing.T) {
+	t.Parallel()
+
+	Test(t, TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"test": providerserver.NewProviderServer(testprovider.Provider{}),
+		},
+		Steps: []TestStep{
+			{
+				Config: `
+				output "list_of_strings" {
+					value = tolist(["hello", "world"])
+				}
+				output "list_of_bools" {
+					value = tolist([true, false])
+				}
+				output "list_of_numbers" {
+					value = tolist([1.23, 4.56, 500])
+				}
+				output "list_of_objects" {
+					value = tolist([{a = "hey", b = "there"}, {a = "and", b = "another"}])
+				}
+				output "list_of_lists" {
+					value = tolist([tolist(["hey", "there"]), tolist(["and", "another"])])
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"list_of_strings",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.StringExact("hello"),
+								knownvalue.StringExact("world"),
+							},
+						),
+					),
+					statecheck.ExpectKnownOutputValue(
+						"list_of_bools",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.Bool(true),
+								knownvalue.Bool(false),
+							},
+						),
+					),
+					statecheck.ExpectKnownOutputValue(
+						"list_of_numbers",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.Float64Exact(1.23),
+								knownvalue.Float64Exact(4.56),
+								knownvalue.Float64Exact(500),
+							},
+						),
+					),
+					statecheck.ExpectKnownOutputValue(
+						"list_of_objects",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.StringExact("hey"),
+									"b": knownvalue.StringExact("there"),
+								}),
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.StringExact("and"),
+									"b": knownvalue.StringExact("another"),
+								}),
+							},
+						),
+					),
+					statecheck.ExpectKnownOutputValue(
+						"list_of_lists",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hey"),
+										knownvalue.StringExact("there"),
+									},
+								),
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("and"),
+										knownvalue.StringExact("another"),
+									},
+								),
+							},
+						),
+					),
+				},
+			},
+		},
+	})
+}
+
+// Ref: https://github.com/hashicorp/terraform-plugin-testing/issues/310
+func TestStateShimOutput_Tuple(t *testing.T) {
+	t.Parallel()
+
+	Test(t, TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"test": providerserver.NewProviderServer(testprovider.Provider{}),
+		},
+		Steps: []TestStep{
+			{
+				Config: `output "test" {
+					value = [true, "hello", 1.23, ["hello", "world"], {a = false}]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.Bool(true),
+								knownvalue.StringExact("hello"),
+								knownvalue.Float64Exact(1.23),
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hello"),
+										knownvalue.StringExact("world"),
+									},
+								),
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.Bool(false),
+								}),
+							},
+						),
+					),
+				},
+			},
+			{
+				Config: `output "test" {
+					value = [{a = false}, true, "hello", 1.23, ["hello", "world"]]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.Bool(false),
+								}),
+								knownvalue.Bool(true),
+								knownvalue.StringExact("hello"),
+								knownvalue.Float64Exact(1.23),
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hello"),
+										knownvalue.StringExact("world"),
+									},
+								),
+							},
+						),
+					),
+				},
+			},
+			{
+				Config: `output "test" {
+					value = [["hello", "world"], {a = false}, true, "hello", 1.23]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hello"),
+										knownvalue.StringExact("world"),
+									},
+								),
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.Bool(false),
+								}),
+								knownvalue.Bool(true),
+								knownvalue.StringExact("hello"),
+								knownvalue.Float64Exact(1.23),
+							},
+						),
+					),
+				},
+			},
+			{
+				Config: `output "test" {
+					value = [1.23, ["hello", "world"], {a = false}, true, "hello"]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.Float64Exact(1.23),
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hello"),
+										knownvalue.StringExact("world"),
+									},
+								),
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.Bool(false),
+								}),
+								knownvalue.Bool(true),
+								knownvalue.StringExact("hello"),
+							},
+						),
+					),
+				},
+			},
+			{
+				Config: `output "test" {
+					value = ["hello", 1.23, ["hello", "world"], {a = false}, true]
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ListExact(
+							[]knownvalue.Check{
+								knownvalue.StringExact("hello"),
+								knownvalue.Float64Exact(1.23),
+								knownvalue.ListExact(
+									[]knownvalue.Check{
+										knownvalue.StringExact("hello"),
+										knownvalue.StringExact("world"),
+									},
+								),
+								knownvalue.ObjectExact(map[string]knownvalue.Check{
+									"a": knownvalue.Bool(false),
+								}),
+								knownvalue.Bool(true),
+							},
+						),
+					),
+				},
+			},
+		},
+	})
+}
+
+func TestStateShimOutput_Object(t *testing.T) {
+	t.Parallel()
+
+	Test(t, TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"test": providerserver.NewProviderServer(testprovider.Provider{}),
+		},
+		Steps: []TestStep{
+			{
+				Config: `output "test" {
+					value = {
+						a = "hey",
+						b = "there",
+						c = true,
+						d = 1.23,
+					}
+				}`,
+				ConfigStateChecks: []statecheck.StateCheck{
+					statecheck.ExpectKnownOutputValue(
+						"test",
+						knownvalue.ObjectExact(map[string]knownvalue.Check{
+							"a": knownvalue.StringExact("hey"),
+							"b": knownvalue.StringExact("there"),
+							"c": knownvalue.Bool(true),
+							"d": knownvalue.Float64Exact(1.23),
+						}),
+					),
+				},
+			},
+		},
+	})
+}

--- a/helper/resource/testing_new_test.go
+++ b/helper/resource/testing_new_test.go
@@ -141,6 +141,21 @@ func TestShimState(t *testing.T) {
       "list_of_bool": {
         "sensitive": false,
         "value": [true, false, true]
+      },
+      "tuple_one": {
+        "sensitive": false,
+        "value": [true, 1.23, "hello world"],
+		"type": ["tuple", ["bool", "number", "string"]]
+      },
+      "tuple_two": {
+        "sensitive": false,
+        "value": ["hello world", true, 1.23],
+		"type": ["tuple", ["string", "bool", "number"]]
+      },
+      "tuple_three": {
+        "sensitive": false,
+        "value": [1.23, "hello world", true],
+		"type": ["tuple", ["number", "string", "bool"]]
       }
     },
     "root_module": {}
@@ -195,6 +210,27 @@ func TestShimState(t *testing.T) {
 								Type: "list",
 								Value: []interface{}{
 									true, false, true,
+								},
+								Sensitive: false,
+							},
+							"tuple_one": {
+								Type: "list",
+								Value: []interface{}{
+									true, json.Number("1.23"), "hello world",
+								},
+								Sensitive: false,
+							},
+							"tuple_two": {
+								Type: "list",
+								Value: []interface{}{
+									"hello world", true, json.Number("1.23"),
+								},
+								Sensitive: false,
+							},
+							"tuple_three": {
+								Type: "list",
+								Value: []interface{}{
+									json.Number("1.23"), "hello world", true,
 								},
 								Sensitive: false,
 							},

